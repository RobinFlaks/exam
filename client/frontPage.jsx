import {Link} from "react-router-dom";
import * as React from "react";

export function FrontPage() {

    return <div>
        <h2>hello!</h2>
        <div>
            <Link to={{pathname: "login"}}>login?</Link>
        </div>
    </div>
}