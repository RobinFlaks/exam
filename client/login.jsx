import * as React from "react";

export function Login(){
    return(
        <div>
            <h3>this is the Login page :D</h3>
        </div>
    )
}