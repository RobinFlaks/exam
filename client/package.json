{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "scripts": {
    "build": "parcel build index.html",
    "dev": "parcel watch index.html",
    "test": "jest"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "babel-jest": "^29.3.1",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "jsdom": "^20.0.3",
    "parcel": "^2.8.0",
    "process": "^0.11.10"
  },
  "dependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.4.3"
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      "@babel/preset-react"
    ]
  },
  "jest": {
    "testEnvironment": "jsdom",
    "globals": {
      "IS_REACT_ACT_ENVIRONMENT": true
    },
    "collectCoverageFrom": [
      "*.jsx"
    ]
  }
}
