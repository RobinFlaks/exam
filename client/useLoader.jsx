import { useEffect, useState } from "react";

export async function fetchJSON(url, options = {}) {
  const res = await fetch(url, {
    method: options.method || "get",
    headers: options.json ? { "content-type": "application/json" } : {},
    body: options.json && JSON.stringify(options.json),
  });

  if (!res.ok) {
    throw new Error(`loading error: ${res.status} -> ${res.statusText};`);
  }

  if (res.status === 200) {
    return await res.json();
  }
}

export function useLoader(loadingFn) {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState();
  const [data, setData] = useState();

  useEffect(() => {
    const loadData = async () => {
      setLoading(true);
      try {
        setData(await loadingFn());
      } catch (error) {
        setError(error);
      }
    };
    loadData().then((r) => setLoading(false));
  }, []);

  return { loading, error, data };
}
