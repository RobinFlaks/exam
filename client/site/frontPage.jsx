import { Link } from "react-router-dom";
import * as React from "react";
import { useLoader } from "../useLoader";

export function FrontPage({ userApi }) {
  const { loading, error, data } = useLoader(async () => userApi.returnUser());

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <div>
      <h2>welcome {data ? data[0].username : "guest"} </h2>
      <div>
        <Link to={{ pathname: "menu" }}>our menu</Link>
      </div>
    </div>
  );
}
