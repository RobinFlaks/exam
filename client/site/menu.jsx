import { useLoader } from "../useLoader";

function ListDishes({menuApi}) {
  const { loading, error, data } = useLoader(async () =>
        menuApi.listMenu()
  );

  if (loading) {
    return <div>loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h2> error: {error.toString()}</h2>
      </div>
    );
  }

  return (
    <div>
      {data.map((dish) => (
        <div>
          <h2>{dish.name}</h2>
          <h4>a short description of the dish: {dish.description}</h4>
          <h5>ingredients: {dish.ingredients}</h5>
          <h5>allergens: {dish.allergens}</h5>
        </div>
      ))}
    </div>
  );
}

export function Menu({menuApi}) {
  return (
    <div>
      <h3>Our dishes:</h3>
      <ListDishes menuApi={menuApi}/>
    </div>
  );
}
