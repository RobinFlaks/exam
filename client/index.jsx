import * as React from "react";
import { createRoot } from "react-dom/client";
import { BrowserRouter, Route, Routes } from "react-router-dom";
import { FrontPage } from "./site/frontPage";
import { Login } from "./banner/login";
import { Register } from "./banner/register";
import { Menu } from "./site/menu";
import { Banner } from "./banner/banner";
import { fetchJSON } from "./useLoader";
import { Logout } from "./banner/logout";

const element = document.getElementById("app");

const root = createRoot(element);

function Application() {
  const menuApi = {
    async listMenu() {
      return fetchJSON("api/menu");
    },
  };
  const userApi = {
    async returnUser() {
      return fetchJSON("api/users");
    },

    async registerUser(user) {
      await fetch("/api/users", {
        method: "post",
        body: JSON.stringify(user),
        headers: {
          "content-type": "application/json",
        },
      });
    },

    async login(user) {
      await fetch("/api/users/login", {
        method: "post",
        body: JSON.stringify(user),
        headers: {
          "content-type": "application/json",
        },
      });
    },

    async logout() {
      await fetch("/api/users/logout", {
        method: "post",
        headers: {
          "content-type": "application/json",
        },
      });
    },
  };
  return (
    <BrowserRouter>
      <Banner userApi={userApi} />
      <Routes>
        <Route path="/" element={<FrontPage userApi={userApi} />} />
        <Route path="/login" element={<Login userApi={userApi} />} />
        <Route path="/register" element={<Register userApi={userApi} />} />
        <Route path="/menu" element={<Menu menuApi={menuApi} />} />
        <Route path="/logout" element={<Logout userApi={userApi} />} />
      </Routes>
    </BrowserRouter>
  );
}

root.render(<Application />);
