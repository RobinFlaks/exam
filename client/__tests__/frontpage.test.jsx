import * as React from "react";
import { createRoot } from "react-dom/client";
import { act } from "react-dom/test-utils";
import { FrontPage } from "../site/frontPage";
import { MemoryRouter } from "react-router-dom";

describe("frontpage test suite", () => {
  it("frontpage is served", () => {
    const element = document.createElement("div");

    const root = createRoot(element);

    act(() => {
      root.render(
        <MemoryRouter>
          <FrontPage />
        </MemoryRouter>
      );
    });

    expect(element.querySelector("div")?.innerHTML).toEqual("Loading...");
    expect(element.innerHTML).toMatchSnapshot();
  });
});
