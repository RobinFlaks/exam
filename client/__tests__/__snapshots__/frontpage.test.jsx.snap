// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frontpage test suite frontpage is served 1`] = `"<div>Loading...</div>"`;
